### ISSUE TITLE
TLS session resumption by default in urllib3

### ISSUE THREAD

**Issue by PleasantMachine9** [urllib3/urllib3#1898]
@PleasantMachine9 Before we start writing code can we discuss within an issue what TLS session resumption by default in urllib3 would look like (where do tickets live, under which conditions do we attempt to resume sessions, what exact performance gains are we achieving by adding this support) and once we nail that down we can start a PR that accomplishes that.

_Originally posted by @sethmlarson in https://github.com/urllib3/urllib3/pull/1886#issuecomment-651308799_

**Comment by PleasantMachine9** [comment]
Some short initial comments below. It's a bit disorganised becuase it's hard to collect all details.

In the current impl. proposal, the TLS sessions would live on the HTTPSConnection level, not pool level. This would mean each connection in the pool would have its own SSLSession reference (or None). I think this is most in line logically with the concept of a TLS "session". 

Sharing it between connections could yield better performance in specific cases, but I think it would make the implementation way more complex (for example, one-time use session tickets exist and those would not be poolable, and this info is not exposed on the python API of openssl).

## where do tickets live
Few things to clarify:

> **(TLSv1.3)**
> 2.2.  Resumption and Pre-Shared Key (PSK)
>
>   Although TLS PSKs can be established out of band, PSKs can also be
>   established in a previous connection and then used to establish a new
>   connection ("session resumption" or "resuming" with a PSK).  Once a
>   handshake has completed, the server can send the client a PSK
>   identity that corresponds to a unique key derived from the initial
>   handshake (see Section 4.6.1).  The client can then use that PSK
>   identity in future handshakes to negotiate the use of the associated
>   PSK.  If the server accepts the PSK, then the security context of the
>   new connection is cryptographically tied to the original connection
>   and the key derived from the initial handshake is used to bootstrap
>   the cryptographic state instead of a full handshake.  In TLS 1.2 and
>   below, this functionality was provided by "session IDs" and "session
>   tickets" [RFC5077].  **Both mechanisms are obsoleted in TLS 1.3.**
>
>   PSKs can be used with (EC)DHE key exchange in order to provide
>   forward secrecy in combination with shared keys, or can be used
>   alone, at the cost of losing forward secrecy for the application
>   data.

- (C)Python's ssl library is a wrapper around the openssl project, which is written in C. The two C structs of interest are wrapped by python's `SSLContext` and `SSLSocket`.

- `SSLSocket.session` is a shallow wrapper around an `SSLSession` struct from openssl.  In the python docs, only a few key attributes are exposed from the underlying C struct: `id, time, timeout, ticket_lifetime_hint, has_ticket`

- Note that as outlined above, from python we cannot actually access the ticket or the _real_ session id bytestrings. The `id` attribute is not the TLS session id, but something else (its meaning is undocumented in the python docs).

- From the openssl codebase, it seems each `SSLSocket` contains a globally unique reference to an `SSLSession` struct. This struct contains various data, and when applicable, contains the references to session ticket and id bytestrings, after the handshake is completed (but they are inaccessible from python).

- When a session is "reused" by a subsequent connection, on the openssl level, a new SSLSession struct will be created from it, even if the resumption is successful (`is` operator check always evaluates to `False`). The python attr. `SSLSocket.session_reused` can be used to actually check the openssl layer on whether the resumption succeeded.

I did not research much further than this into openssl, not sure if that's scope here, and I don't have deep understanding of C.

What's clear is that this solution of sessions is very much bound to the openssl wrapper provided with CPython. So openssl cannot be swapped out, unless a wrapper with the same attrs. and behavior replaces it.

## under which conditions do we attempt to resume sessions

To clarify, the word "attempt" here means that
> "we pass the previous `SSLSession` reference to `wrap_socket` when creating a new socket".

We are not going to be doing the bulk of the work. If the session resumption fails for any reason (too old session, server does not accept, invalid session)  openssl seems to transparently fall back to doing a normal handshake, we can only tell whether resumption happened via the attr. `SSLSocket.session_reused` on the new connection.

Currently the simple solution I see is:

- if resumption is enabled (by default?) and the appropriate attributes (`SSLSocket.session`, modern `SSLContext`) exist at runtime, we always "attempt" it
- otherwise we set OP_NO_TICKET if possible, and do not "attempt" it (by not saving the reference to the `SSLSocket.session`)


##  what exact performance gains are we achieving by adding this support

The goal is the ability to save the time and computation needed by SSL handshakes which can be expensive on bad quality connections and/or low-power devices. Basically similar reasons to why browsers support this feature.

My specific use-case for the inception of this was a livestreaming server which for some reason was not handling `Connection: keep-alive` properly, and so `requests` (which uses `urrllib3`) was re-establishing a new TCP _and_ a new TLS connection every 2 seconds. Since a TLS handshake is 3-4 RTT (if the certs are big it takes 1-2 more packets), that increases the RTT of the new connection (as the TCP handshake is 3).

The 0-RTT fature is also possible *in the TLS spec*, but I don't think the python impl. would allow us to use it even if it were possible with openssl. For that we would already need to pass early data in the `wrap_socket` call (as the handshake happens there). Anyway, 0-RTT is very problematic due to replay attacks.

Prerequisite:
- A previously opened `HTTPSConnection` was closed on the TCP layer, and is being reused by the `HTTPSConnectionPool` containing it

On TLSv1.2, the realistic savings in a happy case are:

- reduce the RTT for the connection setup (6-7 packets to 5)
- save >=2KB of data transmission on wire (rough estimate of >=2KB server certs + OCSP optionally + misc handshake data like client/server randoms)
- save some computation time (resumption skips ECDHE KEX and RSA/ECDSA server/client auth steps)

Cons:
- risk of compromising forward secrecy, if the server is using tickets (ie. not ids), and the server is not rotating or keeping their ticketing keys correctly (note that *before* this change we are already receiving these tickets on the wire from some servers, as OP_NO_TICKET is off by default, we are just not using them, but this already compromises forward secrecy)

On TLSv1.3, it gets more complicated:

> PSKs can be used **with (EC)DHE key exchange** in order to provide forward secrecy

- RTT is only reduced significantly if ECDHE KEX is skipped, ie. client sends the PSK in initial ClientHello. The current openssl behavior seems to be sending the PSK (ticket) in ClientHello initially but this is also hard to tell because of how TLSv1.3 pretends to be 1.2. 
- certificates are *probably not sent* if resumption is done, but this is hard to tell because the handshake is not in the clear in TLSv1.3 anymore
- if ECDHE KEX was not skipped, then we are not saving that computation. RSA/ECDSA server/client auth is probably skipped with a PSK, so some computation may still be skipped.

**Comment by sethmlarson** [comment]
Okay, first of all thank you for writing up all of this and doing tons of research. I apologize that I've taken a long time on getting an official response to this (started writing one up some time ago but must not have been magically saved by GitHub)

Here are my thoughts:
- Should we be setting `OP_NO_TICKET` on all connections in v1.25.x (cc @tiran) to mitigate servers sending us tickets we drop on the floor?
- 100% agree it's much easier to track and implement TLS resumption on the HTTPSConnection level rather than pool even if pool would result in better performance for small numbers of resumptions per connection.
- There seems to be a lot of literature out there about how TLS 1.2 session tickets are tough to get right and/or flat out insecure?
  Obviously there are performance improvements to be had and I wouldn't want it on by default but could see an option to enable it if it's done right? Probably more learning to be done here by looking at other HTTP clients and browsers too.
- I would like to do more research on TLSv1.3 resumption and decide if there's something for us to do there. Are there plans on the horizon to support this option in CPython? (also cc @tiran)
  Quote from CPython docs:
  > SSLSocket.session and SSLSession are not compatible with TLS 1.3.
- Since this work involves a feature only available in Python 3.6+ what are your thoughts to adding it on the v2.0 (Python 3.6+ already) feature list so it's simpler to implement?

**Comment by tiran** [comment]
TLS session resumption is tricky to use with TLS 1.2. IIRC clients should only use a ticket when the previous connection was closed and finalized without any protocol error. A ticket must only be used once and not shared across SSL_CTXs. For TLS 1.3 it's even more complex and not yet implemented in Python's ssl module. Tickets are send out of bounce after the handshake, a server sends multiple tickets, and the session API requires callbacks.

I recommend that you ignore tickets for now. Persistent connections are more efficient.

**Comment by PleasantMachine9** [comment]
### Should we be setting OP_NO_TICKET on all connections in v1.25.x (cc @tiran) to mitigate servers sending us tickets we drop on the floor?

It makes no sense to allow tickets in transport unless we actually ever plan to use them.
It's worth noting that on TLS1.3 the OP_NO_TICKET doesn't seem to do anything in particular, but python doesn't really do much with TLS1.3 anyway.

### 100% agree it's much easier to track and implement TLS resumption on the HTTPSConnection level rather than pool

I was toying around with the idea. If a session can be "resumed" while it's still ongoing, this could make pools quite a bit more performant, when using multiple sockets. There may not need to be a handshake.
But according to tiran, I guess the concept of session here means only 1 socket connection should be active at once. If so, then it only makes sense on HTTPSConnection level, not on Pool level.

### IIRC clients should only use a ticket when the previous connection was closed and finalized without any protocol error.

Regarding this comment, I could not find such explicit information in the RFC about session tickets:
https://tools.ietf.org/html/rfc5077

The only mention that may be related is here:

>    The TLS specification requires that TLS sessions be invalidated when
>    errors occur.  [CSSC] discusses the security implications of this in
>    detail.  **In the analysis within this paper, failure to invalidate**
>    **sessions does not pose a security risk.**  This is because the TLS
>    handshake uses a non-reversible function to derive keys for a session
>    so information about one session does not provide an advantage to
>    attack the master secret or a different session.  If a session
>    invalidation scheme is used, the implementation should verify the
>    integrity of the ticket before using the contents to invalidate a
>    session to ensure that an attacker cannot invalidate a chosen
>    session.

Overall, considering than python's `ssl` is just a wrapper around openssl, if this must be done then I would expect openssl to do it, not `urllib3`.

To begin with, from the available python bindings, we cannot see whether the session was resumed or a full handshake was performed either, except for the `session_resumed` attribute on the SSLSocket _after_ the handshake.

### not shared across SSL_CTXs.

Not sure what term you mean here by CTX, the implementation, or is this from the RFC?
Looking at the openssl interface, the `SSLSocket.session` attribute actually [does contain a reference to its originating context](https://github.com/python/cpython/blob/8394500cca56490cc347604d39ca40abcdce46c3/Modules/_ssl.c#L484) (although not visible from python), so I assume this is ensured by openssl itself.

### I recommend that you ignore tickets for now. Persistent connections are more efficient.

Please check the PR for the rationale. Some HTTPS servers do not always honor `Connection: keep-alive`, and in those cases, this is still better than a full handshake repeat every 2-4 seconds (for example).

I'm trying to replicate browser HTTP client behavior, because Firefox and Chrome do use tickets, and as a result have significantly better latency than `requests` in some specific use-cases.

**Comment by tiran** [comment]
The majority of servers at least supports TLS 1.3. There is no OpenSSL, BoringSSL or NSS version without TLS 1.3 support that is supported by upstream.

OpenSSL 1.1.0+ invalidates session tickets when the underlying TCP connection is closed without performing at least a one-way TLS shutdown. I have added a hack to the ssl module because Python's ssl module does not support one-way shutdown and HTTP clients typically don't perform a two-way shutdown. High level consumers must track the origin of a session ticket and not reuse a ticket when the source connection is in an error state. https://github.com/openssl/openssl/issues/1550 contains more information

``SSL_CTX`` is an OpenSSL struct. ``ssl.SSLContext`` is basically a wrapper for ``SSL_CTX``. Consumers cannot take a session object and reuse it with a different context object, https://www.openssl.org/docs/man1.1.0/man3/SSL_set_session.html .

**Comment by PleasantMachine9** [comment]
> The majority of servers at least supports TLS 1.3. There is no OpenSSL, BoringSSL or NSS version without TLS 1.3 support that is supported by upstream.

I see two problems with this though. The first is that python's own ssl binding is not supporting TLSv1.3 fully at this point, and a lot of servers are still using TLSv1.2, even _if_ their software may support 1.3 (ie. it is not enabled).

The 2019 closing share of browser(?) connections using TLSv1.3 [was just under 30% over the board](https://www.ietf.org/blog/tls13-adoption/), and I am guessing most of those browsers supported 1.3, so it was the servers that did not have it enabled. From my anecdotal experience 1.3 adoption (where it is actually _enabled_) is pretty low compared to what it could be, and I don't see a ton of servers changing unless there is a critical issue with 1.2...

Actually the specific HLS streaming service I use also does not have TLS 1.3 enabled, so that wouldn't have been useful for my use case anyway. I guess I could live with applying the patches locally for the time, but I just don't see a clear timeline where TLSv1.3 is adopted fully by _both_ the HLS server in question _and python3_.

> SSL_CTX is an OpenSSL struct. ssl.SSLContext is basically a wrapper for SSL_CTX. Consumers cannot take a session object and reuse it with a different context object, https://www.openssl.org/docs/man1.1.0/man3/SSL_set_session.html 

Ok, so as for the PR, not only the `SSLSocket.session` but also the `SSLSocket.context` should be "saved" into instance-level variables when resumption is enabled, and both should be passed to `wrap_socket` when reponening the connection. That should be a trivial fix.

**Comment by PleasantMachine9** [comment]
One additional note: I was wrong, the HLS server I use does support TLSv1.3, but python3.8+ssl does not, at least not by default.

**Comment by tiran** [comment]
> I see two problems with this though. The first is that python's own ssl binding is not supporting TLSv1.3 fully at this point, and a lot of servers are still using TLSv1.2, even if their software may support 1.3 (ie. it is not enabled).

The ssl module supports TLS 1.3 fully excepts for 0-RTT and session resumption.

Applications and libraries should try to reuse an SSLContext object when possible. It's inefficient to re-create a context. It takes easily 3-10ms or more depending on settings, type and amount of root CA certificates.

**Comment by PleasantMachine9** [comment]
The SSLContext objects are saved between reconnections already in the current `urllib3` impl, based on the code (on HTTPSConnection level, not pool level).

On TLSv1.3, you are right, it looks like `urllib3` is sending `supported_versions` extension with TLSv1.3, so in this case I guess it is the server that does not support it, and always falls back to picking TLSv1.2 (these are AWS servers).

One interesing note is that `TLSv1.0` and `TLSv1.1` is also in the supported versions sent by `urllib3`, I guess for legacy reasons. Current browsers negotiate strictly v1.2 or 1.3. Maybe we should also set ` ssl.OP_NO_TLSv1` and ` ssl.OP_NO_TLSv1_1` by default? This is unrelated to this issue though.

**Comment by tiran** [comment]
Applications and libraries should neither define their own crypto policy (TLS versions and cipher suite) nor should they ship with their own set of root CA certificates. Instead they should use the platform's crypto settings and default CA store.

**Comment by PleasantMachine9** [comment]
So to sum up current points about the original topic


> TLS session resumption is tricky to use with TLS 1.2. IIRC clients should only use a ticket when the previous connection was closed and finalized without any protocol error.

- [ ] Not sure if this is done by OpenSSL or not.

It looks like urllib3 does not care/track this. When attempting to reuse a socket, it just uses some arcane magic code to check if the connection is dropped (unwriteable socket (?)), and unilaterally closes the socket if it deems so, in practice via TCP RST according to pcap logs.

I did not produce any tests locally when there were any "protocol errors". It seems like in practice this would be a really difficult situation to produce naturally so not sure how to even test it.

> High level consumers must track the origin of a session ticket and not reuse a ticket when the source connection is in an error state.

- [ ] Not sure what this means in practice either for urllib3.

I did not dig deep into the socket interfacing modules because they seem very convoluted to me (probably due to cross-platform scenarios), but it seems like there is no logic tracking any socket error states, and actually it does not seem like these errors are even exposed by python's `ssl` module after the initial handshake(?).
The only logic I could find is checking whether connections are "dropped" or not.

> A ticket must only be used once and not shared across SSL_CTXs.

- [x] Already done this way in the PR.

> For TLS 1.3 it's even more complex and not yet implemented in Python's ssl module. 

- [x] The PR mainly actually targets TLSv1.2.

A lot of web servers on the internet, including the one in my use-case are running on TLSv1.2-only. So I don't see how TLSv1.3 not being implemented in python's `ssl` is an issue. That just means it will continue to not work (as it doesn't work now) on TLSv1.3, no?

I also responded above to the share of servers which actually have TLSv1.3 _enabled_ and it's not as high by far as the number of servers which would support it on paper. 

### Summary

I guess the part where the decision should be made is

- is there any logic needed to detect graceful vs. ungraceful shutdowns of the SSL sockets - arguable, in the RFC I quoted above, if I am reading it right, it is not considered a security issue, even if tickets are not invalidated:

>  In the analysis within this paper, failure to invalidate
sessions does not pose a security risk.

- if detection of ungraceful socket shutdown is needed, how/where to do it. I am lost on this one because the socket-related low-level code sections are pretty unclear to me

**Comment by eighthave** [comment]
Another detail to keep in mind: TLS Sessions can be used like a cookie.  In TLSv1.2, that cookie is plain text, so any network that the packet passes through can see it.  It is encrypted in TLSv1.3, but the server can still use it as a cookie.

* https://www.theregister.com/2018/10/19/tls_handshake_privacy/
* https://gitlab.torproject.org/legacy/trac/-/issues/4099

**Comment by sethmlarson** [comment]
I don't think we can do this in urllib3, and I'm not sure it matters much since TLS 1.3 is becoming the default for many servers.
